src/goto-programs/add_race_assertions.cpp

```diff
@@ -37,20 +41,35 @@ class w_guardst
     return *symbol_ptr;
   }
 
-  const exprt get_guard_symbol_expr(const irep_idt &object)
+  const exprt
+  get_guard_symbol_expr(const irep_idt &object, const exprt &original_expr)
   {
-    return symbol_expr(get_guard_symbol(object));
+    exprt expr = symbol_expr(get_guard_symbol(object, original_expr));
+
+    if(original_expr.is_index() && expr.type().is_array())
```

Comment:
What about if we have a struct instead of an array?

Reply:
This type depends on the `tmp` symbol we create, so far we only create two types: `array` and `bool`, we treat sruct/member as "id +'.' + component.as_string()" bool type symbol. 

https://github.com/esbmc/esbmc/blob/0afdaa2400136ef78567f6092e63509e1c8e5c60/src/goto-programs/rw_set.cpp#L68-L73

We'll consider these types later if needed.